Add ui in studio for conditional_module.

oksana-slu opened the pull request:
Background: Usage and modification of Conditional Module can be undertaken only with course import/export mechanism. Now Creation and modification of Conditional Mode are enabled within Studio functionality.

Studio Updates: The next file is modified: common/lib/xmodule/xmodule/conditional_module.py . For new functionality, the next tests are corrected: common/lib/xmodule/xmodule/tests/test_conditional.py; the new tests are added.

LMS Updates: None

Testing:

![img 1](https://dl.dropboxusercontent.com/u/28163465/conditional_1.png)

![img 2](https://dl.dropboxusercontent.com/u/28163465/conditional_2.png)

![img 3](https://dl.dropboxusercontent.com/u/28163465/conditional_3.png)

![img 4](https://dl.dropboxusercontent.com/u/28163465/conditional_4.png)

![img 5](https://dl.dropboxusercontent.com/u/28163465/conditional_5.png)

![img 6](https://dl.dropboxusercontent.com/u/28163465/conditional_6.png)

![img 7](https://dl.dropboxusercontent.com/u/28163465/conditional_7.png)

edx-webhook commented:
Thanks for the pull request, @oksana-slu! I've created [OSPR-1177](https://openedx.atlassian.net/browse/OSPR-1177) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

AlesyaD commented:
@jbarciauskas

ormsbee commented:
Thanks for the PR! I just took a quick glance at it, and the approach looks good to me (simple, and doesn't impact our longer term plans to make this an XBlock). I'm not really familiar with this module, but my only concern would be around data migration -- making sure that old courses still import correctly with the new fields, and that there aren't odd ways to create legitimate conditional blocks that would break your UI.

nedbat commented:
jenkins ok to test

nedbat commented:
@scottrish product review?

oksana-slu commented:
@justinabrahms @ormsbee 
We have fixed a tests, so all tests finished successfully . Also, we tested this features with the old courses ( Export / import). Old courses still import correctly with the new fields

jbarciauskas commented:
@oksana-slu This will need UX review as well as regular engineering review. I've created a second ticket in JIRA, you can find it there.

oksana-slu commented:
@jbarciauskas 
we have sandbox with this feature, please follow the link to make UX review
LMS: http://178.62.209.246/
CMS: http://178.62.209.246:18010/

marcotuts commented:
Hi! Thank you for the sandbox setup here and the details, I'm excited for what this will enable for our course teams in Studio! I would love to set up some time to review this in more detail to make sure I understand all the configuration options and the workflow for authors. Would it be possible to find some time to meet virtually? If it helps with scheduling we can use calendly.com/marcotuts

Thanks!
Marco

oksana-slu commented:
@marcotuts 
Hi, we are glad that you find this feature useful for your course team. Of course, we can meet virtually.
Olesya booked 21 April at 4:15PM via calendly.com/marcotuts. Or we can chat anytime.

jbarciauskas commented:
@marcotuts did you get together to review this?

marcotuts commented:
@jbarciauskas We did meet, and THIS is the PR that I owe a product review / response. I will get to that before Monday for sure, apologies for the delay on this @oksana-slu! 

Thanks

nedbat commented:
@marcotuts ping

nedbat commented:
@marcotuts we need to do something with this PR.  If we aren't going to be able to respond, we should say that.

marcotuts commented:
Here is a summary of the product feedback so far. 

# The use of the A/B testing / Randomized Problem Block pattern of nesting the components inside of the block is great! 
# A bit more effort is necessary on naming and wording in a number of places. I've suggested a list of areas where these changes could happen, but @catong may have other suggestions as well. I've also slightly reordered the settings as well. 

## Setting name: Source Components --> originally "sources list"
### The component location(s) whose attributes are used to determine whether a student should be shown the content inside of this conditional block.
## Setting: Conditional Attribute --> (originally "conditional_attr")
### Setting subtitle - The attribute from the source component used to determine whether students should be shown the content in this conditional block.
## Setting: Conditional Value --> (originally "conditional_value")
### Setting subtitle - The value of the conditional attribute that needs to be true for a given student to see the content in this conditional block.
## Setting name: Blocked content message --> (originally "message")
### Setting subtitle - The message learners see when not all conditions are met for this block. You can use the {link} variable to give learners a direct link to the required module.

## Warning message on component: This component has no source components configured yet. 

FYI's
@andy-armstrong 
@felipemontoya - who has worked on an XBlock similar to this.

marcotuts commented:
@felipemontoya I know this PR still needs to undergo a review, but I wanted to call out the Flow control XBlock work you did and see if it would be possible spark some discussions with @oksana-slu and team in the future (after this PR, more than likely) about how certain conditional block fields might be improved. As an example, the flow control XBlock's message area being a WYSIWYG area is interesting.

oksana-slu commented:
@marcotuts 
Hello,
we would like to clarify about PRs.
Paragraph 1-7 - everything is clear and we fix it asap.
In paragraph 8 you say about location attribute. But, everything what concerns location we set in PR Location (https://github.com/edx/edx-platform/pull/11999).
Should we merge them?

marcotuts commented:
@oksana-slu Sorry, I can move my comment for 8 to the other PR, you are correct that it mostly concerns this other PR's effort. Thanks!

oksana-slu commented:
@marcotuts 
We updated the sandbox with conditional module
LMS: http://edx-sendr-dogwood.raccoongang.com/
CMS: http://studio-sendr-dogwood.raccoongang.com/

jbarciauskas commented:
Note to whoever picks this up: please coordinate with @pdpinch and MIT to do some manual testing

marcotuts commented:
Also tagging @Colin-Fredericks here as an FYI for this module.

Colin-Fredericks commented:
Thanks, @marcotuts . Stupid question from the guy who just got added to the thread: Do these nest?

shaunagm commented:
@marcotuts @Colin-Fredericks What's the status on this PR?  What needs to happen next?

marcotuts commented:
@shaunagm - thanks for checking in! I believe this is on the review queue for the teaching and learning team. cc'ed @andy-armstrong, @dianakhuang  - This would be good to make sure we get into the #6 sprint. Thanks everyone!

jbarciauskas commented:
bump @marcotuts @shaunagm @gsong

marcotuts commented:
Hi! This OSPR review made it into our current sprint (2 weeks) that opened today. I expect either @andy-armstrong or another team member from T&L will be reviewing this. Thanks!

andy-armstrong commented:
@oksana-slu I will be reviewing this PR this week. Can you please rebase so that we can try to get a  clean build for it. Also, it looks like the sandboxes need to be regenerated. In the meantime, I'll start looking at the code.

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -10,20 +10,68 @@
+    )
+
+    show_tag_list = ReferenceList(
+        help=_("List of urls of children that are references to external modules"),
```
@catong could you review the help strings here when you have a chance. Thanks.

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -10,20 +10,68 @@
+        scope=Scope.content
+    )
+
+    condional_attr = String(
```
Typo: this should be ``conditional_attr``.

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -95,27 +143,15 @@ class ConditionalModule(ConditionalFields, XModule):
 
     def is_condition_satisfied(self):
-        xml_value, attr_name = self._get_condition()
+        attr_name = self.conditions_map[self.condional_attr]
```
The specified attribute might not be in the map, since it is a user supplied value. Should this code be defensive against that situation, or is that not possible?

andy-armstrong replied:
Nit: I'd move this line inside the ``if`` since it isn't used elsewhere.

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -147,18 +183,27 @@ def get_html(self):
             'depends': ';'.join(self.required_html_ids)
         })
 
+    def author_view(self, context):
```
Do you need to also override ``has_author_view`` to have this take effect?

oksana-slu replied:
This  has_author_view has value = true by default because it is identified in class https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/studio_editable.py#L13

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -147,18 +183,27 @@ def get_html(self):
+        fragment = Fragment()
+        root_xblock = context.get('root_xblock')
+        is_root = root_xblock and root_xblock.location == self.location
+        if is_root:
```
Should this render something if the block is not a root? It seems odd that it would just return an empty fragment.

andy-armstrong replied:
I see that this is a common convention, so the block just shows its message. Maybe that warrants a comment to make this clear.

oksana-slu replied:
We will add comments.

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -231,6 +285,14 @@ def get_required_module_descriptors(self):
+            conditional_value = xml_object.get(condional_attr)
+            if conditional_value is not None:
+                definition.update({
+                    'condional_attr': condional_attr,
```
Typo: ``'condional_attr'``

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -262,4 +328,36 @@ def definition_to_xml(self, resource_fs):
+            conditional_validation.add(
+                StudioValidationMessage(
+                    StudioValidationMessage.NOT_CONFIGURED,
+                    _(u"This component has no source components configured yet."),
```
@catong any suggestions for this message?

catong replied:
I'm fine with this existing message. I think it's consistent with others in the platform, e.g. content libraries.

catong commented:
@andy-armstrong Thanks for the tag. I hope to get to UI string review later today.

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -10,20 +10,68 @@
+        default='True'
+    )
+
+    conditional_message = String(
```
Do we need to support a way to localize the messages? It seems odd that the default message is localized, but the author's message is not.

oksana-slu replied:
I'm not sure about a general way of solving this issue. According to the source code of Open edX (https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/course_module.py#L421), this string is localised.
Could you please give an advice what should I do in my case.

catong commented:
@andy-armstrong @oksana-slu FYI I have also reviewed the related https://github.com/edx/edx-platform/pull/11999.

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -231,6 +285,14 @@ def get_required_module_descriptors(self):
         children = []
         show_tag_list = []
+        definition = {}
+        for condional_attr in ConditionalModule.conditions_map.iterkeys():
```
Typo: ``condional_attr``

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -231,6 +285,14 @@ def get_required_module_descriptors(self):
         show_tag_list = []
+        definition = {}
+        for condional_attr in ConditionalModule.conditions_map.iterkeys():
+            conditional_value = xml_object.get(condional_attr)
```
Should this have a default value?

oksana-slu replied:
This conditional_value is identified on the class level

andy-armstrong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -95,27 +143,15 @@ class ConditionalModule(ConditionalFields, XModule):
         'voted': 'voted'  # poll_question attr
     }
 
-    def _get_condition(self):
```
I'm nervous about the change in functionality for pre-existing conditional modules. It seems that the current behavior is to find the first conditional attribute, but now it will always look for an attribute called ``correct``. It might be safer to have ``conditional_attr`` be optional, and if it isn't specified then the current behavior is maintained. Since all pre-existing modules won't have this field set, then they will continue to behave as before.

andy-armstrong replied:
Thinking this through some more, the logic would have to ignore the specified ``conditional_value`` field in this case too, as that won't match for pre-existing courses.

oksana-slu replied:
Yes, I see  this is a real problem for described case. 
Would it be right decision to remove default value for the following fields 'conditional_attr', 'conditional_value' and implement this functionality the same way as in (https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/conditional_module.py#L202)

andy-armstrong commented:
@oksana-slu This feature is excellent! I love how you reused the ability to have container pages. I've finished my first pass code review, so let me know when you're ready for another review (and when the sandbox is available).

oksana-slu commented:
@andy-armstrong 
Thank you for reviewing.
I will try to fix everything till the next Monday

oksana-slu commented:
@andy-armstrong 
I added all comment, looking forward to your feedback

andy-armstrong commented:
@oksana-slu can you rebase this again and let me know when the tests are passing. I'll re-review once there is a green build. Thanks.

catong commented:
@andy-armstrong @oksana-slu Is our current pattern to present the configuration fields alphabetically? I believe @marcotuts suggested some reordering in [this comment](https://github.com/edx/edx-platform/pull/11710#issuecomment-226019042) and I agree that changing the order of some of the fields would make the settings much more understandable to users. 
If you agree that they don't have to be presented alphabetically I'll include suggestions for revised order as part of my help strings review.

catong commented:
@oksana-slu To expedite, here are my edited field labels and Help text, presented in what seems to me a logical order. The conditional module's own non-editable Location ID comes last. @marcotuts or @andy-armstrong please comment. My only unknown is the syntax for the ```{link}``` variable - I'd like to include an example in the Help text (see the ```XXXX``` placeholder below).

**Display Name**
This name appears in the horizontal navigation at the top of the page.

**Source Components**
The locations of the components whose attributes are used to determine whether a learner is shown the content of this conditional module.

**Conditional Attribute**
The attribute from the course component used to determine whether a learner is shown the content of this conditional module.

**Conditional Value**
The value of the conditional attribute that must be true for a learner to be shown the content of this conditional module.

**Blocked Content Message**
The message that is shown to learners when not all conditions are met to show the content of this conditional module. To include a link to the required module within the message, use the required module's location ID with the {link} variable. For example, XXXX

Location
The location ID of this conditional module.

marcotuts commented:
@catojg - Sorry just wanted to chime in here - I don't think we should reorder the fields in this Pr and we should generally speaking retain alphabetical order. The component location may have been a separate PR (I didn't remember the ordering comment I made. I was later convinced t should be on all components and not a Buried / hidden field. As much as I don't like alphabetical ordering of the fields I actually think the experience is harder to parse and find fields while authoring if they are organized any other way.

catong commented:
@marcotuts That's fine - I reordered the fields in my comment above into alphabetical order. Thanks for commenting!

oksana-slu commented:
@andy-armstrong
I fixed the rebase issue.
We updated the sandbox
LMS: https://edx-ui-conditional-module.raccoongang.com
Studio: https://studio-ui-conditional-module.raccoongang.com
Could you please help us, we have problems with tests.
this one - test_conditional_handles_polls failed in master too.
Tests lettuce type finished local correctly but failed in jenkins.

catong commented:
@marcotuts, on further discussion with @andy-armstrong and @robrap, it seems that not all XBlocks do present settings fields in alphabetical order, e.g. LTI. I think it would contribute a great deal to usability if we arranged the fields in more logical order. It's especially distracting to have the first field be "Blocked Content Message". Can we reconsider?

catong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -10,20 +10,68 @@
+
+    sources_list = ReferenceList(
+        display_name=_("Source Components"),
+        help=_("The location IDs of the components whose attributes are used to determine whether a learner is shown "
```
@oksana-slu @andy-armstrong Do I understand correctly that more than one source component can be specified? How does a course team member enter more than one component? (e.g. separated by commas?)

marcotuts commented:
@oksana-slu - I wanted to flag that the current sandbox doesn't seem to include the component location PR as a dependency, meaning it isn't possible to configure new conditionals. For the time being, anybody who needs to test a conditional block can use this component source: 

block-v1:raccoon+CN+09_2016+type@problem+block@385e66d65b2a48f28930395c3ef31644

No need to change sandbox for now I think, as this component ID unblocks testing, but I wanted to make sure I shared this in the PR. Thanks!

catong commented:
@marcotuts Per our conversation I have tweaked the Help text and field labels. Please review and let us know if OK. Thanks!
**Blocked Content Message**
The message that is shown to learners when not all conditions are met to show the content of this conditional module. Include ```{link}``` in the text of your message to give learners a direct link to required units. For example, "You must complete {link} before you can access this unit".

**Conditional Attribute**
The attribute of the source components that determines whether a learner is shown the content of this conditional module.

**Conditional Value**
The value that the conditional attribute of the source components must match before a learner is shown the content of this conditional module.

**Display Name**
This name appears in the horizontal navigation at the top of the page.

**Component Location ID**
The component location ID of this conditional module.

**Source Components**
The component location IDs of all source components that are used to determine whether a learner is shown the content of this conditional module. Copy the component location ID of a component from its Settings dialog in Studio.

catong commented on common/lib/xmodule/xmodule/conditional_module.py:
```diff
@@ -10,20 +10,68 @@
+        help=_("The message learners see when not all conditions are met for this block. "
+               "You can use the {link} variable to give learners a direct link to the required module."),
+        scope=Scope.content,
+        default=_('{link} must be attempted before this will become visible.')
```
@oksana-slu Please update this default text to read:
```You must complete {link} before you can access this unit.```

marcotuts commented:
Thanks @catong! All those suggestions make sense to me.

catong commented:
@oksana-slu Can you please implement that Help and Field Label text changes? Thank you!

marcotuts commented:
jenkins run bokchoy
jenkins run lettuce

marcotuts commented:
@oksana-slu - Any sense of when you will be able to get to the conditional poll bokchoy test failure? After this @andy-armstrong can continue his review.

gsong commented:
@marcotuts [See @oksana-slu's comment](issuecomment-245301113) regarding getting help about that test specifically. Any idea who can help with that from our end?

marcotuts commented:
@andy-armstrong - Are there any confluence pages / suggestions we could link @oksana-slu to re: common issues with: "local test success / jenkins test fail" situations. I know it came up in conversation that it might be a situation of needing some sort of wait() or time related delay? I couldn't find the correct Confluence resource myself.

andy-armstrong commented:
@oksana-slu @marcotuts here's a wiki with some tips for Bok Choy:

https://openedx.atlassian.net/wiki/pages/viewpage.action?pageId=39583991

oksana-slu commented:
@marcotuts @andy-armstrong 
I will try to fix this test next week, and inform you about progress.
Thanks

oksana-slu commented:
@andy-armstrong @marcotuts 
I have already fixed the test.  You can continue to review the pull request.
Looking forward to your feedback.

andy-armstrong commented:
@oksana-slu That's great news that you got all the tests to pass! I'll be able to re-review this on Monday. Could you check that the sandbox is up-to-date before I start. Also, did you address the label changes that @catong recommended?

oksana-slu commented:
@andy-armstrong 
Hello, 
the sandbox is updated.
Follow please the link to test it.
And I have also mada changes that you recommended.
LMS: http://edx-ui-conditional-module.raccoongang.com
Studio: http://studio-ui-conditional-module.raccoongang.com

login: staff@example.com
password: edxPass123

andy-armstrong commented:
@oksana-slu for some reason, I'm not seeing the "Conditional Location ID" on then sandbox for any components that I add to the course. That makes it hard to use the feature (although I can work around it). Does it need rebasing to pick up that feature?

andy-armstrong commented:
@oksana-slu the feature works great! I have a few pieces of feedback on the UI but nothing that cannot be treated as follow-on stories just so you can merge all your hard work here.

1. The UI is quite complicated to understand, so I think it would be good to have documentation on the right-hand side like we do for content experiments. Maybe @catong could suggest some options. Here's how content experiments looks:

![image](https://cloud.githubusercontent.com/assets/5985072/19043150/f691e280-895c-11e6-8299-7e1540edd29a.png)

You can see a live component in my test course on your sandbox:

http://studio-ui-conditional-module.raccoongang.com/container/block-v1:raccoongang+AA101+1+type@split_test+block@51f733a536cf41c6b38df1274e71b6d0

2. I think it would also be good to show the components that are influencing the content on the Conditional UI page itself. As it stands, it isn't clear precisely when this content will be shown. I would model it on how the content experiment page shows a label above the content, i.e the part that says: "This content experiment uses group configuration 'Andy's Test Experiment'".

3. This is pre-existing behavior, but it is unfortunate that you can't use a conditional on the same page as the question it is conditional on. If you do (as in my test course), the conditional content doesn't appear when the condition becomes true. Instead, you have to refresh the page to get it to show up. Maybe this is just something that needs to be documented.

I'm going to review the code now.

andy-armstrong commented:
@catong can you re-review when you have a chance to verify that the strings are all as you had in  mind. Thanks!

Colin-Fredericks commented:
It looks like it takes multiple "source" components and applies the same
logic to all of them. Is that the intent?

And yes, some documentation will definitely be necessary, if nothing else
to explain what the different options are for "Conditional Value" and what
the options for "Conditional Attribute" mean. It may be more helpful to
give that inline underneath the attribute than in the sidebar as shown for
Content Experiments, since getting back to that sidebar requires cancelling
out of what you're doing.

Very glad to have this functionality moving into Studio.

On Mon, Oct 3, 2016 at 12:15 PM, Andy Armstrong <notifications@github.com>
wrote:

> @catong <https://github.com/catong> can you re-review when you have a
> chance to verify that the strings are all as you had in mind. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-platform/pull/11710#issuecomment-251150948>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB1xJG8BbkMoKnagJ4Yc2ltEH1uyEp9Bks5qwSongaJpZM4HmcpG>
> .
>

andy-armstrong commented:
@Colin-Fredericks Thanks for the feedback, Colin. I agree that we need a variety of documentation to help folks with this feature. Note that there is some help text beneath each attribute in the settings modal:

![image](https://cloud.githubusercontent.com/assets/5985072/19046604/261a22a6-896c-11e6-855d-977639ea32b3.png)

I'd still like to see documentation on the page detailing what to do, because it starts out as a blank page with a single edit button. 

![image](https://cloud.githubusercontent.com/assets/5985072/19046733/b90c048a-896c-11e6-946a-400c546c6e04.png)

We also need online documentation if authors are going to be able to work out what to do.

andy-armstrong commented:
@catong do you have an ETA for a review of this PR? The documentation changes you requested have all been made, so my preference would be to merge this and have a follow-up story to add some inline help text. Thanks!

catong commented:
@andy-armstrong I agree that improvements to the inline help can follow, but the accompanying course author documentation is still in progress. I think it is Product's (@sstack22) call whether to get this feature out on prod without accompanying doc.

catong commented:
Sorry - I guess it was @marcotuts who was involved with this PR.

andy-armstrong commented:
@catong Note that the RC has been cut for next week's release, so the feature won't be live until the following week at the earliest. I also think that it is okay if the documentation isn't released at the same time, since this is an esoteric feature that must be enabled through an advanced setting, so folks won't be affected by it unless they've been following this work. 

@sstack22, are you comfortable giving thumbs in @marcotuts's absence, or shall we wait for him to return? Thanks.

catong commented:
@andy-armstrong Thanks for clarifying the release timing. I thought this RC had not yet been cut today and did want @sstack22's confirmation.

sstack22 commented:
@andy-armstrong - I can definitely review, but most likely will be able to on Tuesday. Let me know if this is urgent though!

marcotuts commented:
cc'ed @andy-armstrong, @catong, @sstack22 - Ok to merge this in as in I think. A separate follow-on story for in-context documentation can be added to the backlog, and any announcement of this UI in studio can be gated until we have the base documentation. Thoughts on that? 

(Excited to see this nearing merge!)

andy-armstrong commented:
@oksana-slu @marcotuts @catong I've created a new JIRA story to add some inline documentation to this view:

https://openedx.atlassian.net/browse/TNL-5732

I'm going to go ahead and merge this PR now. Thanks for the great work here, @oksana-slu!
